File: ruby/lib/epostak/resources/sapi.rb
# frozen_string_literal: true

require "erb"

module EPostak
  module Resources
    # SAPI-SK 1.0 interoperable document send/receive endpoints.
    class Sapi
      def initialize(http, base_url:)
        @http = http
        @base_url = base_url
        @participants = SapiParticipants.new(self)
      end

      attr_reader :participants

      def send_document(body, participant_id:, idempotency_key:)
        sapi_request(
          :post,
          "/sapi/v1/document/send",
          body: body,
          participant_id: participant_id,
          idempotency_key: idempotency_key
        )
      end

      def receive(participant_id:, limit: nil, status: nil, page_token: nil)
        query = compact(limit: limit, status: status, pageToken: page_token)
        sapi_request(
          :get,
          "/sapi/v1/document/receive#{query_string(query)}",
          participant_id: participant_id
        )
      end

      def get(document_id, participant_id:)
        sapi_request(
          :get,
          "/sapi/v1/document/receive/#{encode(document_id)}",
          participant_id: participant_id
        )
      end

      def acknowledge(document_id, participant_id:)
        sapi_request(
          :post,
          "/sapi/v1/document/receive/#{encode(document_id)}/acknowledge",
          participant_id: participant_id
        )
      end

      private

      def sapi_request(method, path, participant_id:, body: nil, idempotency_key: nil)
        @http.request(
          method,
          path,
          body: body,
          idempotency_key: idempotency_key,
          query: nil,
          retry_on_failure: method == :get,
          headers: { "X-Peppol-Participant-Id" => participant_id },
          base_url: @base_url.sub(%r{/api/v1/?\z}, "")
        )
      end

      def compact(hash)
        hash.reject { |_k, v| v.nil? }
      end

      def query_string(hash)
        return "" if hash.empty?

        "?" + hash.map { |k, v| "#{encode(k)}=#{encode(v)}" }.join("&")
      end

      def encode(value)
        ERB::Util.url_encode(value)
      end
    end

    class SapiParticipants
      def initialize(sapi)
        @sapi = sapi
      end

      def for_participant(participant_id)
        SapiParticipant.new(@sapi, participant_id)
      end
    end

    class SapiParticipant
      attr_reader :documents

      def initialize(sapi, participant_id)
        @documents = SapiParticipantDocuments.new(sapi, participant_id)
      end
    end

    class SapiParticipantDocuments
      def initialize(sapi, participant_id)
        @sapi = sapi
        @participant_id = participant_id
      end

      def send_document(body, idempotency_key:)
        @sapi.send_document(body, participant_id: @participant_id, idempotency_key: idempotency_key)
      end

      def send(body, idempotency_key:)
        send_document(body, idempotency_key: idempotency_key)
      end

      def receive(limit: nil, status: nil, page_token: nil)
        @sapi.receive(participant_id: @participant_id, limit: limit, status: status, page_token: page_token)
      end

      def get(document_id)
        @sapi.get(document_id, participant_id: @participant_id)
      end

      def acknowledge(document_id)
        @sapi.acknowledge(document_id, participant_id: @participant_id)
      end
    end
  end
end

File: python/src/epostak/resources/sapi.py
"""SAPI-SK 1.0 interoperable document resource."""

from __future__ import annotations

import re
from typing import Any, Dict, Optional
from urllib.parse import quote, urlencode

from epostak.resources.documents import _BaseResource, _build_query


class SapiResource(_BaseResource):
    """SAPI-SK 1.0 document send/receive endpoints."""

    @property
    def _sapi_base_url(self) -> str:
        return re.sub(r"/api/v1/?$", "", self._base_url)

    def _sapi_request(
        self,
        method: str,
        path: str,
        *,
        participant_id: str,
        json: Any = None,
        extra_headers: Optional[Dict[str, str]] = None,
    ) -> Any:
        original = self._base_url
        headers = {"X-Peppol-Participant-Id": participant_id}
        if extra_headers:
            headers.update(extra_headers)
        try:
            self._base_url = self._sapi_base_url
            return self._request(method, path, json=json, extra_headers=headers)
        finally:
            self._base_url = original

    def send(
        self,
        body: Dict[str, Any],
        *,
        participant_id: str,
        idempotency_key: str,
    ) -> Dict[str, Any]:
        """Send one SAPI XML document."""
        return self._sapi_request(
            "POST",
            "/sapi/v1/document/send",
            participant_id=participant_id,
            json=body,
            extra_headers={"Idempotency-Key": idempotency_key},
        )

    def receive(
        self,
        *,
        participant_id: str,
        limit: Optional[int] = None,
        status: Optional[str] = None,
        page_token: Optional[str] = None,
    ) -> Dict[str, Any]:
        """List received SAPI documents for a participant."""
        query = _build_query({"limit": limit, "status": status, "pageToken": page_token})
        suffix = f"?{urlencode(query)}" if query else ""
        return self._sapi_request(
            "GET",
            f"/sapi/v1/document/receive{suffix}",
            participant_id=participant_id,
        )

    def get(self, document_id: str, *, participant_id: str) -> Dict[str, Any]:
        """Get a received SAPI document with its XML payload."""
        return self._sapi_request(
            "GET",
            f"/sapi/v1/document/receive/{quote(document_id, safe='')}",
            participant_id=participant_id,
        )

    def acknowledge(self, document_id: str, *, participant_id: str) -> Dict[str, Any]:
        """Mark a received SAPI document as processed."""
        return self._sapi_request(
            "POST",
            f"/sapi/v1/document/receive/{quote(document_id, safe='')}/acknowledge",
            participant_id=participant_id,
        )
